bundles/org.openhab.binding.mercedesme/src/main/java/org/openhab/binding/mercedesme/internal/server/CallbackServer.java

```diff
@@ -0,0 +1,190 @@
+import org.openhab.core.auth.client.oauth2.OAuthFactory;
+import org.openhab.core.auth.client.oauth2.OAuthResponseException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link CallbackServer} class defines an HTTP Server for authentication callbacks
+ *
+ * @author Bernd Weymann - Initial contribution
+ */
+@NonNullByDefault
+public class CallbackServer {
+    private static final Logger LOGGER = LoggerFactory.getLogger(CallbackServer.class);
+    private static final Map<Integer, OAuthClientService> AUTH_MAP = new HashMap<Integer, OAuthClientService>();
+    private static final Map<Integer, CallbackServer> SERVER_MAP = new HashMap<Integer, CallbackServer>();
+
+    private Optional<Server> server = Optional.empty();
+    private Optional<AccessTokenResponse> token = Optional.empty();
+    private AccessTokenRefreshListener listener;
+    private AccountConfiguration config;
+    private OAuthClientService oacs;
+    private String callbackUrl;
+
+    public CallbackServer(AccessTokenRefreshListener l, HttpClient hc, OAuthFactory oAuthFactory,
+            AccountConfiguration config, String callbackUrl) {
+        oacs = oAuthFactory.createOAuthClientService(Constants.OAUTH_CLIENT_NAME + config.callbackPort,
+                Constants.MB_TOKEN_URL, Constants.MB_AUTH_URL, config.clientId, config.clientSecret, config.getScope(),
+                false);
+        AUTH_MAP.put(Integer.valueOf(config.callbackPort), oacs);
+        SERVER_MAP.put(Integer.valueOf(config.callbackPort), this);
+        listener = l;
+        this.config = config;
+        this.callbackUrl = callbackUrl;
+    }
+
+    public String getAuthorizationUrl() {
+        try {
+            return oacs.getAuthorizationUrl(callbackUrl, null, null);
+        } catch (OAuthException e) {
+            LOGGER.error("Error creating Authorization URL {}", e.getMessage());
+            return Constants.EMPTY;
+        }
+    }
+
+    public String getScope() {
+        return config.getScope();
+    }
+
+    public void start() {
+        LOGGER.debug("Start Callback Server for port {}", config.callbackPort);
+        if (!server.isEmpty()) {
+            LOGGER.debug("Callback server for port {} already started", config.callbackPort);
+            return;
+        }
+        server = Optional.of(new Server());
+        ServerConnector connector = new ServerConnector(server.get());
+        connector.setPort(config.callbackPort);
+        server.get().setConnectors(new Connector[] { connector });
+        ServletHandler servletHandler = new ServletHandler();
+        server.get().setHandler(servletHandler);
+        servletHandler.addServletWithMapping(CallbackServlet.class, Constants.CALLBACK_ENDPOINT);
+        try {
+            server.get().start();
+        } catch (Exception e) {
+            LOGGER.warn("Cannot start Callback Server for port {}, Error {}", config.callbackPort, e.getMessage());
```

Comment:
Same here, I would propose throwing an exception so the message could be propagated back to caller and handled, for example by using it for `updateStatus()`.

Reply:
It's also part of the provided rework